reply API 스펙 변경

sharpie1330 opened the pull request:
- year 파라미터 추가, 연도별 필터링
- 응답에 published 필드 추가

sharpie1330 committed:
refactor: 연도별로 편지함 조회 가능하도록 변경, published 필드 추가해 응답하도록 변경

jellyyelly commented on src/main/java/bsise/server/reply/ReplyService.java:
```diff
@@ -90,10 +93,13 @@ public List<ReplyResponseDto> findMyLetterAndReply(UUID userId, Integer size) {
         validateUserId(userId);
 
-        Page<Reply> replies = replyRepository.findRepliesByOrderByCreatedAt(userId, pageable);
+        LocalDateTime startOfYear = Year.of(year).atDay(1).atStartOfDay();
+        LocalDateTime nextStartOfYear = startOfYear.plusYears(1);
+
+        Page<Reply> replies = replyRepository.findRepliesByOrderByCreatedAt(userId, startOfYear, nextStartOfYear, pageable);
```
`year` 값으로 2024를 받으면 `2024-01-01 - 2025-01-01` 사이의 편지들을 조회하는 것으로 오해될 소지가 있는 것 같습니다.
아래 코드로 제안 드려요

```suggestion
        LocalDate startOfYear = Year.of(year).atDay(1);
        LocalDate endOfYear = startOfYear.with(TemporalAdjusters.lastDayOfYear());

        Page<Reply> replies = replyRepository.findRepliesByOrderByCreatedAt(userId, LocalDateTime.of(startOfYear, LocalTime.MIN), LocalDateTime.of(endOfYear, LocalTime.MAX), pageable);
```

sharpie1330 replied:
음... 원래 제안주신 코드대로 구현했다가, LocalTime.MAX와 `<=` 비교를 해도 혹시나 23:59:59.999999... 이시점에 작성된 게 포함 안되면 어쩌지 하는 생각이 들어서 jpql의 between을 대신해 `createdAt >= startOfYear AND createdAt < nextStartOfYear` 이렇게 변경했는데요.

사실 저는 오해의 소지가 그렇게 있나? 싶기는 해요.
하지만 다른 부분 코드와 통일성이 있어야 할 것 같기는 해서... 변경해서 커밋해두겠습니다

jellyyelly replied:
`nextStartOfYear` 변수명이 "어떤 연도의 다음 시작일(?)"이라고 받아들여졌습니다
차라리 `startOfNextYear` 이면 "다음 연도의 시작일(1월 1일)"가 나을 수 있겠네요..!


\+  `startOfNextYear`라고 해도 다음 해 1월 1일까지 데이터를 가져오는 것 같아서 `endOfYear`로 맞췄습니다!

\+ `'23:59:59.999999...' 이시점에 작성된..` ⇒ MySQL에서 created_at 필드는 `datetime(6)` 타입으로 저장되서 소수점 6자리를 초과해서 저장될 일이 없습니다~! 이 부분은 걱정 안하셔도 될 것 같아요 ㅎㅎ (feat.RealMySQL 8.0 2권 데이터 타입)

jellyyelly replied:
근데 제가 예전에 만든 LetterRepository에도 조건이 샤피랑 똑같네요 ;;
과거의 나는 왜그랬을까..


아.. 제가 아니라 그로밋이었네요..ㅋㅋㅋㅋㅋ

sharpie1330 replied:
ㅋㅋㅋ 수정 중입니다~ 완료하고 바로 PR 할게요!

jellyyelly commented on src/main/java/bsise/server/reply/ReplyRepository.java:
```diff
@@ -25,6 +27,6 @@ public interface ReplyRepository extends JpaRepository<Reply, UUID> {
 
-    @Query("SELECT r FROM Reply r where r.letter.user.id = :userId ORDER BY r.createdAt DESC")
-    Page<Reply> findRepliesByOrderByCreatedAt(UUID userId, Pageable pageable);
+    @Query("SELECT r FROM Reply r where r.letter.user.id = :userId AND r.createdAt >= :startOfYear AND r.createdAt < :nextStartOfYear ORDER BY r.createdAt DESC")
+    Page<Reply> findRepliesByOrderByCreatedAt(UUID userId, LocalDateTime startOfYear, LocalDateTime nextStartOfYear, Pageable pageable);
```
아래 제안한 코드에 따라 이렇게 변경하는걸 제안 드려요

```suggestion
    @Query("SELECT r FROM Reply r where r.letter.user.id = :userId AND r.createdAt >= :startOfYear AND r.createdAt <= :endOfYear ORDER BY r.createdAt DESC")
    Page<Reply> findRepliesByOrderByCreatedAt(UUID userId, LocalDateTime startOfYear, LocalDateTime endOfYear, Pageable pageable);
```

jellyyelly commented on src/main/java/bsise/server/reply/ReplyController.java:
```diff
@@ -44,14 +44,15 @@ public List<ReplyResponseDto> getTopNLetters(
     @ResponseStatus(HttpStatus.OK)
     public Page<ReplyResponseDto> findMyLetters(
             @PathVariable("userId") String userId,
+            @RequestParam int year,
```
변수명이 변경되어도 프론트 단에서 변경이 없도록 이름을 지정하면 어떨까요?
```suggestion
            @RequestParam("year") int year,
```

sharpie1330 replied:
오 변수명 변경될 것이라는 생각을 못했군요
감사합니다~

jellyyelly reviewed (commented):
제안드린 코드 검토 부탁해요~!

sharpie1330 committed:
refactor: 연도별 reply 조회 리팩토링

- requestParam 변수 변경 시 대처 가능하도록 name 설정
- nextStartOfYear 대신 endOfYear로 BETWEEN 범위 조회하도록 범위설정 변경해 다른 코드와 통일성 유지
